mito-ai/mito_ai/utils/mito_server_utils.py

```diff
@@ -88,5 +89,147 @@ async def get_response_from_mito_server(
+        chunk_processor: Optional function to process chunks before yielding (e.g., for Gemini's special processing)
+        
+    Yields:
+        Chunks of text from the streaming response
+    """
+    # Check the mito server quota
+    check_mito_server_quota(message_type)
+    
+    # Create HTTP client with appropriate timeout settings
+    http_client, http_client_timeout = _create_http_client(timeout, max_retries)
+    
+    # Set up streaming infrastructure
+    start_time = time.time()
+    chunk_queue: asyncio.Queue[str] = asyncio.Queue()
+    fetch_complete = False
+    
+    # Define a callback to process chunks and add them to the queue
+    def chunk_callback(chunk: bytes) -> None:
+        try:
+            chunk_str = chunk.decode('utf-8')
+            asyncio.create_task(chunk_queue.put(chunk_str))
+        except Exception as e:
+            print(f"Error processing {provider_name} streaming chunk: {str(e)}")
+    
+    # Execute the streaming request
+    fetch_future = None
+    try:
+        fetch_future = http_client.fetch(
+            url, 
+            method="POST", 
+            headers=headers, 
+            body=json.dumps(data), 
+            request_timeout=http_client_timeout,
+            streaming_callback=chunk_callback
+        )
+        
+        # Create a task to wait for the fetch to complete
+        async def wait_for_fetch() -> None:
+            try:
+                await fetch_future
+                nonlocal fetch_complete
+                fetch_complete = True
+                print(f"{provider_name} fetch completed")
+            except Exception as e:
+                print(f"Error in {provider_name} fetch: {str(e)}")
+                raise
+        
+        # Start the task to wait for fetch completion
+        fetch_task = asyncio.create_task(wait_for_fetch())
+        
+        # Yield chunks as they arrive
+        while not (fetch_complete and chunk_queue.empty()):
+            try:
+                # Wait for a chunk with a timeout to prevent deadlocks
+                chunk = await asyncio.wait_for(chunk_queue.get(), timeout=0.1)
+                
+                # Process chunk if processor is provided
+                processed_chunk = chunk
+                if chunk_processor:
+                    processed_chunk = chunk_processor(chunk)
+                
+                # If reply_fn is provided, send the chunk directly to the frontend
+                if reply_fn and message_id:
+                    reply_fn(CompletionStreamChunk(
+                        parent_id=message_id,
+                        chunk=CompletionItem(
+                            content=processed_chunk,
+                            isIncomplete=True,
+                            token=message_id,
+                        ),
+                        done=False,
+                    ))
+                
+                yield chunk  # Yield original chunk for backward compatibility
```

Comment:
I think this comment is wrong. The reason we are yielding the chunks is so that the backend can collect the full Ai response and save it the chat-history.json file.